### PR TITLE
add k_usleep() API for microsecond-resolution sleeps

### PR DESCRIPTION
**Pull request by ghost**
These patches add k_usleep() support on x86 with the HPET timer, and an associated test.

The local APIC should also be perfectly serviceable for k_usleep(), but there currrently problems when trying to compile a kernel that uses LOAPIC_TIMER and TICKLESS; I will open a bug for this and try to root-cause it.

There are no limitations in these changes that limit ~~it~~ them to x86 only, but not every board will provide the infrastructure required to support fine-resolution sleeps.  I thought it better to adopt a more conservative "opt-in" approach, exposing the API only to known-good targets, rather than giving users a function that might or might not work.

### CODE REVIEW

File: kernel/Kconfig [hunk 1]
```diff
@@ -760,6 +760,21 @@ config TICKLESS_KERNEL
 	  This option enables a fully event driven kernel. Periodic system
 	  clock interrupt generation would be stopped at all times.
 
+config USLEEP
```
Comment (andyross) [inline kernel/Kconfig]: I think you can skip this kconfig.  There's no cost if this is enabled but unused, and it doesn't change the internals of anything.  It's just an extra API.  The linker will drop it for us.

File: tests/kernel/usleep/src/main.c [hunk 2]
```diff
@@ -0,0 +1,45 @@
...
+ */
+
+#include <ztest.h>
+#include <zephyr.h>
+
+#define SLEEP_US 	50
+#define LOOPS 		200
+
+#define LOWER_BOUND_MS	((SLEEP_US * LOOPS) / 1000)
+#define UPPER_BOUND_MS	(LOWER_BOUND_MS * 2)
```
Comment (andyross) [inline tests/kernel/usleep/src/main.c]: Seems like a factor of two is more than you need.  The lower bound is safe, but this one could be tightened quite a bit.  By my read the test below should be accurate to +/- 2 ticks.

**Review by andyross — APPROVED** [review]
Looks great on the whole.  Nitpicks about the need for a separate kconfig and the tolerance values in the test.

File: kernel/Kconfig (hunk 1 above)
Comment (gnuless) [reply kernel/Kconfig]: TIL that linkers are way smarter than they used to be.  Hmph.

File: tests/kernel/usleep/src/main.c [hunk 3]
```diff
@@ -0,0 +1,67 @@
...
+	while (retries < RETRIES) {
+		s64_t start_ms;
+		s64_t end_ms;
+		int i;
+
+		++retries;
+
+		start_ms = k_uptime_get();
+
+		for (i = 0; i < LOOPS; ++i)
```
Comment (jukkar) [inline tests/kernel/usleep/src/main.c]: Minor style issue here, I wonder why checkpatch does not complain about this. Anyway, you need to add {} to the one liners in this this function (here and couple of lines below).

**Comment by nashif** [comment]
> There are no limitations in these changes that limit ~it~ them to x86 only, but not every board will provide the infrastructure required to support fine-resolution sleeps.

what timers we have in zephyr do not support this?

File: tests/kernel/usleep/src/main.c (hunk 3 above)
Comment (gnuless) [reply tests/kernel/usleep/src/main.c]: My understanding is that the Linux style guidelines (that checkpatch is meant to enforce) specifically decry the use of braces around simple one-liners:

(http://lxr.linux.no/linux+v3.6.1/Documentation/CodingStyle#L157)

But the braces are required by MISRA, right?

**Comment by gnuless** [comment]
> what timers we have in zephyr that do not support this?

Well, the local APIC timer is one, although that's not a hardware limitation, we just need to untangle the TICKLESS from the legacy TICKLESS_IDLE stuff.

Most (if not all) the other timers will probably support some degree of k_usleep().  The problem is that use of k_usleep() will not work out of the box: it requires tweaking some other configuration values, like the number of ticks per second and basically requiring (for sane operation) a TICKLESS kernel.  Messing with those options can introduce other problems.  For example, for 10us sleep resolution, ticks must be no longer than 10us (-> ticks/sec is 100,000), and since internal kernel timers work on ticks, and tick counts are s32_t, timing suddenly (silently) breaks when dealing with intervals greater than a few hours.  We can't simple bump the size of the tick count without universally penalizing everyone, even those without a need for higher resolution timers.

Some platforms can't even get 10us resolution.  Some of the NRFs I think used fixed clocks at 32.768kHz, meaning ~ 30uS is the best that can be done. 

I decided the best option was to hide k_usleep() behind a config option, so that it's not presented to, and assumed to be functional, by the average Zephyr programmer who is unaware of these subtleties.  Plus, the configuration option may be used at some point to do some intelligent adaptation in the timer code (e.g., use 64-bit tick counts).  There is other work planned to rethink/rework the kernel timer code to address some of these things, which is why I don't attempt to address them in this patch.

File: kernel/Kconfig (hunk 1 above)
Comment (gnuless) [reply kernel/Kconfig]: There's another reason for the config option though.  See my comment to @nashif below.  The "other work planned" is, I believe, something you've got at least a hand in. :)

File: tests/kernel/usleep/src/main.c (hunk 2 above)
Comment (gnuless) [reply tests/kernel/usleep/src/main.c]: I agree, but I can't get the test past QEMU reliably without the wider margin.  Real hardware, no problem.  Emulation.. not so much.

Comment (andyross) [reply tests/kernel/usleep/src/main.c]: Qemu is a problem regardless.  This doesn't look to me like it's particularly "more" susceptible to the host interactions than other tests.    But sure.  We can always tighten things later.

**Comment by gnuless** [comment]
I seem to have screwed up this branch somehow.  Closing and will open a new pull request.

**Comment by jukkar** [comment]
> I seem to have screwed up this branch somehow. Closing and will open a new pull request.

Was there something that force pushing would not have solved? Now all the comments are kind of "lost".

File: tests/kernel/usleep/src/main.c (hunk 3 above)
Comment (jukkar) [reply tests/kernel/usleep/src/main.c]: I thought that checkpatch would have checked the braces but actually it does not do it.
And yes, it is some MISRA rule.

**Pull request closed without merging**